### PR TITLE
Hotfix/revoke refresh access token

### PR DESCRIPTION
**Pull request by Asheshp23**
* Threw the error from the methods( Fetch data, Uplate data, logout) so that SSSDK utilizer can catch the error and logout the user if the refresh access token is expired.

### CODE REVIEW

File: Sources/SimpleSalesforceSDK/Models/SFAuth.swift [hunk 1]
```diff
@@ -57,8 +57,13 @@ class SFAuth {
...
   func refreshAccessTokenIfNeeded(config: SFConfig) async throws {
     guard !self.isAccessTokenValid else { return }
-    
-    try! await self.refreshAccessToken(config: config)
+    do {
+      try await self.refreshAccessToken(config: config)
+    } catch {
+      self.reset()
+      print("Error while refreshing the access token...")
+      print(String(describing: error))
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/SFAuth.swift]: Don’t you want to raise again? So the user of the SDK knows that something went wrong?

File: Sources/SimpleSalesforceSDK/Models/SFAuth.swift [hunk 2]
```diff
@@ -103,12 +114,16 @@ class SFAuth {
...
-    self.accessTokenExpiryDate = expiryDate
+
+    do {
+      let (data, _) = try await WebService.makeRequest(request)
+      let responseData = try! JSONDecoder().decode(IntrospectResponse.self, from: data)
+
+      let expiryDate = Date(timeIntervalSince1970: TimeInterval(responseData.accessTokenExpiryDate))
+      self.accessTokenExpiryDate = expiryDate
+    } catch {
+      throw error
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/SFAuth.swift]: Don’t need a try catch, if you’re just throwing it back

File: Sources/SimpleSalesforceSDK/Models/SFAuth.swift [hunk 3]
```diff
@@ -123,6 +138,10 @@ class SFAuth {
     let requestConfig = RequestConfig(url: url, params: params.data(using: .utf8))
     let request = URLRequestBuilder.request(with: requestConfig)
 
-    let _ = try! await WebService.makeRequest(request)
+    do {
+      let _ = try await WebService.makeRequest(request)
+    } catch {
+      throw error
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/SFAuth.swift]: Same as above

File: Sources/SimpleSalesforceSDK/Models/SFAuth.swift [hunk 4]
```diff
@@ -57,8 +57,7 @@ class SFAuth {
   /// This will also call `interospectAccessToken` to reset the expiry
   func refreshAccessTokenIfNeeded(config: SFConfig) async throws {
     guard !self.isAccessTokenValid else { return }
-    
-    try! await self.refreshAccessToken(config: config)
+      try await self.refreshAccessToken(config: config)
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/SFAuth.swift]: Why is the indentation changed?

File: Sources/SimpleSalesforceSDK/Models/SFAuth.swift [hunk 5]
```diff
@@ -79,13 +78,19 @@ class SFAuth {
...
+      decoder.dateDecodingStrategy = .secondsSince1970
+      let responseData = try! decoder.decode(RefreshTokenResponse.self, from: data)
 
-    self.accessToken = responseData.accessToken
-    try! await self.interospectAccessToken(config: config)
+      self.accessToken = responseData.accessToken
+      try! await self.interospectAccessToken(config: config)
+    } catch {
+      self.reset()
+      print("Error while refreshing the access token...")
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/SFAuth.swift]: Why aren’t we raising the error again after resetting the auth? How will the caller of the method know something went wrong?

File: Sources/SimpleSalesforceSDK/Models/SFAuth.swift [hunk 6]
```diff
@@ -123,6 +128,6 @@ class SFAuth {
     let requestConfig = RequestConfig(url: url, params: params.data(using: .utf8))
     let request = URLRequestBuilder.request(with: requestConfig)
 
-    let _ = try! await WebService.makeRequest(request)
+      let _ = try await WebService.makeRequest(request)
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/SFAuth.swift]: Why change in indentation?

File: Sources/SimpleSalesforceSDK/Models/SFAuth.swift [hunk 7]
```diff
@@ -74,16 +73,14 @@ class SFAuth {
     "&client_id=\(config.clientId)" +
     "&refresh_token=\(refreshToken)" +
     "&client_secret=\(config.clientSecret)"
-
+
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/SFAuth.swift]: why do we need this change?

File: Sources/SimpleSalesforceSDK/Models/SFAuth.swift [hunk 8]
```diff
@@ -95,34 +92,30 @@ class SFAuth {
   /// Including whether this token is currently active, expiry, originally issued
   func interospectAccessToken(config: SFConfig) async throws {
     let url = try! URLBuilder.introspectURL(urlString: config.host)
-
+
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/SFAuth.swift]: why do we need this change?

File: Sources/SimpleSalesforceSDK/Models/SFAuth.swift [hunk 9]
```diff
@@ -95,34 +92,30 @@ class SFAuth {
...
   func interospectAccessToken(config: SFConfig) async throws {
     let url = try! URLBuilder.introspectURL(urlString: config.host)
-
+    
     let params = "token=\(self.accessToken!)" +
     "&client_id=\(config.clientId)" +
     "&client_secret=\(config.clientSecret)" +
     "&token_type_hint=access_token"
-
+
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/SFAuth.swift]: why do we need this change?

File: Sources/SimpleSalesforceSDK/Models/SFAuth.swift [hunk 10]
```diff
@@ -95,34 +92,30 @@ class SFAuth {
...
-    
-    let (data, _) = try! await WebService.makeRequest(request)
+    let (data, _) = try await WebService.makeRequest(request)
     let responseData = try! JSONDecoder().decode(IntrospectResponse.self, from: data)
-    
     let expiryDate = Date(timeIntervalSince1970: TimeInterval(responseData.accessTokenExpiryDate))
     self.accessTokenExpiryDate = expiryDate
   }
-
+
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/SFAuth.swift]: why do we need this change?

File: Sources/SimpleSalesforceSDK/Models/SFAuth.swift [hunk 11]
```diff
@@ -95,34 +92,30 @@ class SFAuth {
...
-
+  
   /// Revokes the salesforce access token.
   /// - Parameters:
   ///     - config: The Salesforce instance’s configuration.
   ///
   /// This method requests salesforce to invalidate and revoke the access token.
   func revokeAccessToken(config: SFConfig) async throws {
     let url = try! URLBuilder.revokeTokenURL(urlString: config.host)
-
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/SFAuth.swift:121]: why do we need this change?

File: Sources/SimpleSalesforceSDK/SSSDK.swift [hunk 12]
```diff
@@ -10,19 +10,19 @@ import SwiftUI
 ///
 /// Once authenticated, you can also use this class to make API calls to Salesforce.
 public class SSSDK {
-
+
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/SSSDK.swift]: why do we need this change?

Comment (Asheshp23) [reply Sources/SimpleSalesforceSDK/SSSDK.swift]: Before it was not indented properly so i re-indented using XCode's feature of re-indenting.

File: Sources/SimpleSalesforceSDK/Models/SFAuth.swift [hunk 13]
```diff
@@ -50,17 +50,14 @@ class SFAuth {
...
-  
   /// Fetches a new access token only if it is already expired and stores it to Keychain
   /// - Parameters:
   ///     - config: The Salesforce instance’s configuration.
   /// This will also call `interospectAccessToken` to reset the expiry
   func refreshAccessTokenIfNeeded(config: SFConfig) async throws {
     guard !self.isAccessTokenValid else { return }
-    
-    try! await self.refreshAccessToken(config: config)
+    try await self.refreshAccessToken(config: config)
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/SFAuth.swift:61]: how is this compiling? https://cocoacasts.com/what-is-the-difference-between-try-try-and-try

Comment (ipranay) [reply Sources/SimpleSalesforceSDK/Models/SFAuth.swift:61]: nvm, I found the answer... looks like if the method has a `throws` in it, you can just use `try` 🎉

Screen Shot 2022-08-05 at 3 54 31 PM

https://stackoverflow.com/questions/32390611/try-try-try-what-s-the-difference-and-when-to-use-each

File: Sources/SimpleSalesforceSDK/Models/SFAuth.swift [hunk 14]
```diff
@@ -69,60 +66,48 @@ class SFAuth {
...
     self.accessTokenExpiryDate = expiryDate
   }
-
   /// Revokes the salesforce access token.
   /// - Parameters:
   ///     - config: The Salesforce instance’s configuration.
   ///
   /// This method requests salesforce to invalidate and revoke the access token.
   func revokeAccessToken(config: SFConfig) async throws {
     let url = try! URLBuilder.revokeTokenURL(urlString: config.host)
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/SFAuth.swift]: Let's remove all `try!` from the code and make sure all methods say throws when we have `try`s.

File: Sources/SimpleSalesforceSDK/Services/WebService.swift [hunk 15]
```diff
@@ -45,53 +40,49 @@ class WebService {
...
-      try! await updateRecord(
+      try await auth.refreshAccessToken(config: config)
+      try await updateRecord(
         config: config,
         auth: auth,
         id: id,
         objectName: objectName,
         fieldUpdates: fieldUpdates,
-        shouldRetry: false // retry only once
+        shouldRetry: true // retry only once
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Services/WebService.swift]: If you make this `true`.. it will keep retrying infinitely. Please revert this change.

File: Sources/SimpleSalesforceSDK/Services/WebService.swift [hunk 16]
```diff
@@ -79,19 +74,24 @@ class WebService {
...
+    let (data, response) = try await URLSession.shared.data(for: request)
+
     guard let httpResponse = response as? HTTPURLResponse else {
       throw SSSDKError.notOk
     }
 
     let statusCode = httpResponse.statusCode
     guard statusCode >= 200 && statusCode < 299 || (ignore401 && statusCode == 401) else {
-      if let jsonString = String(data: data, encoding: .utf8) {
-        print(jsonString)
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Services/WebService.swift:90]: since you removed the printing of the error, how would the caller know if it is not a `authRefreshTokenExpiredError` error?

Comment (ipranay) [reply Sources/SimpleSalesforceSDK/Services/WebService.swift:90]: Maybe you can include text in the `SSSDKError.notOk` error or just print it like before

Comment (Asheshp23) [reply Sources/SimpleSalesforceSDK/Services/WebService.swift:90]: if it is not the authRefreshTokenExpiredError then i am throwing the SSSDKError.notOk error.

**Review by ipranay — APPROVED** [review]

**Comment by Asheshp23** [comment]
Thanks
